File: development/frappe-bench/apps/university_erp/university_erp/university_erp/page/role_dashboard/role_dashboard.py
# Copyright (c) 2026, University ERP and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import nowdate, add_months, getdate, flt


@frappe.whitelist()
def get_role_dashboard_data():
    """Get dashboard data based on user role"""
    user = frappe.session.user
    roles = frappe.get_roles(user)

    if "Student" in roles:
        return get_student_dashboard(user)
    elif "Instructor" in roles:
        return get_faculty_dashboard(user)
    elif "Education Manager" in roles:
        return get_admin_dashboard()
    elif "HR Manager" in roles:
        return get_hr_dashboard()
    elif "Accounts Manager" in roles:
        return get_finance_dashboard()

    return get_default_dashboard()


def get_student_dashboard(user):
    """Student-specific dashboard"""
    student = frappe.db.get_value("Student", {"user": user}, "name")

    if not student:
        return {"type": "student", "error": "Student record not found"}

    student_doc = frappe.get_doc("Student", student)

    # Get current enrollments
    enrollments = frappe.get_all(
        "Course Enrollment",
        filters={"student": student},
        fields=["course", "enrollment_date"],
        limit=10
    )

    # Enrich with course names
    for e in enrollments:
        e["course_name"] = frappe.db.get_value("Course", e["course"], "course_name")

    # Get attendance (last 30 days)
    attendance = frappe.db.sql("""
        SELECT
            COALESCE(SUM(CASE WHEN status = 'Present' THEN 1 ELSE 0 END), 0) as present,
            COUNT(*) as total
        FROM `tabStudent Attendance`
        WHERE student = %s
        AND date >= DATE_SUB(CURDATE(), INTERVAL 30 DAY)
    """, (student,), as_dict=True)[0]

    attendance_rate = 0
    if attendance.get("total"):
        attendance_rate = round(attendance.get("present", 0) / attendance.get("total") * 100, 1)

    # Get pending fees
    pending_fees = frappe.db.sql("""
        SELECT COALESCE(SUM(outstanding_amount), 0) as total
        FROM `tabFees`
        WHERE student = %s AND docstatus = 1 AND outstanding_amount > 0
    """, (student,))[0][0] or 0

    # Get latest results
    latest_results = frappe.db.sql("""
        SELECT
            ar.course,
            c.course_name,
            ar.custom_grade as grade,
            ar.custom_grade_points as grade_points
        FROM `tabAssessment Result` ar
        JOIN `tabCourse` c ON ar.course = c.name
        WHERE ar.student = %s AND ar.docstatus = 1
        ORDER BY ar.creation DESC
        LIMIT 5
    """, (student,), as_dict=True)

    # Get library status
    books_borrowed = frappe.db.sql("""
        SELECT COUNT(*) FROM `tabLibrary Transaction` lt
        WHERE lt.library_member IN (
            SELECT name FROM `tabLibrary Member` WHERE student = %s
        )
        AND lt.type = 'Issue'
        AND lt.docstatus = 1
        AND lt.name NOT IN (
            SELECT issue_reference FROM `tabLibrary Transaction`
            WHERE type = 'Return' AND docstatus = 1 AND issue_reference IS NOT NULL
        )
    """, (student,))[0][0] or 0

    # Get hostel allocation
    hostel_allocation = frappe.db.get_value(
        "Hostel Allocation",
        {"student": student, "status": "Active"},
        ["hostel_room", "hostel_building"],
        as_dict=True
    )

    # Get placement applications
    placement_applications = frappe.db.count(
        "Placement Application",
        {"student": student, "docstatus": ["!=", 2]}
    )

    return {
        "type": "student",
        "student_name": student_doc.student_name,
        "program": student_doc.program,
        "enrollments": enrollments,
        "attendance_rate": attendance_rate,
        "pending_fees": flt(pending_fees),
        "latest_results": latest_results,
        "books_borrowed": books_borrowed,
        "hostel_allocation": hostel_allocation,
        "placement_applications": placement_applications,
        "cgpa": student_doc.get("custom_cgpa") or student_doc.get("cgpa") or 0
    }


def get_faculty_dashboard(user):
    """Faculty-specific dashboard"""
    employee = frappe.db.get_value("Employee", {"user_id": user}, "name")

    if not employee:
        return {"type": "faculty", "error": "Employee record not found"}

    employee_doc = frappe.get_doc("Employee", employee)

    # Get assigned courses
    assignments = frappe.get_all(
        "Teaching Assignment",
        filters={"instructor": employee, "docstatus": 1},
        fields=["name", "course", "course_name", "weekly_hours", "academic_term"]
    )

    total_hours = sum(a.get("weekly_hours") or 0 for a in assignments)

    # Get today's schedule
    today_schedule = []
    day_map = {0: "Monday", 1: "Tuesday", 2: "Wednesday", 3: "Thursday", 4: "Friday", 5: "Saturday", 6: "Sunday"}
    today_day = day_map.get(getdate(nowdate()).weekday())

    for assignment in assignments:
        schedule = frappe.get_all(
            "Teaching Assignment Schedule",
            filters={"parent": assignment["name"], "day": today_day},
            fields=["start_time", "end_time"],
            order_by="start_time"
        )
        for s in schedule:
            s["course"] = assignment["course"]
            s["course_name"] = assignment["course_name"]
            today_schedule.append(s)

    today_schedule.sort(key=lambda x: x["start_time"])

    # Get pending leave applications
    pending_leaves = frappe.db.count(
        "Leave Application",
        {"employee": employee, "status": "Open"}
    )

    # Get leave balance
    leave_balance = frappe.db.sql("""
        SELECT leave_type, total_leaves_allocated, total_leaves_taken
        FROM `tabLeave Allocation`
        WHERE employee = %s
        AND from_date <= %s AND to_date >= %s
    """, (employee, nowdate(), nowdate()), as_dict=True)

    # Get student feedback summary
    feedback = frappe.db.sql("""
        SELECT AVG(sf.overall_rating) as avg_rating, COUNT(*) as count
        FROM `tabStudent Feedback` sf
        JOIN `tabTeaching Assignment` ta ON sf.teaching_assignment = ta.name
        WHERE ta.instructor = %s AND sf.docstatus = 1
    """, (employee,), as_dict=True)[0]

    return {
        "type": "faculty",
        "employee_name": employee_doc.employee_name,
        "department": employee_doc.department,
        "assignments": assignments,
        "total_hours": total_hours,
        "today_schedule": today_schedule,
        "pending_leaves": pending_leaves,
        "leave_balance": leave_balance,
        "avg_feedback": round(feedback.get("avg_rating") or 0, 1),
        "feedback_count": feedback.get("count") or 0
    }


def get_admin_dashboard():
    """Education admin dashboard"""
    return {
        "type": "admin",
        "total_students": frappe.db.count("Student", {"enabled": 1}),
        "total_programs": frappe.db.count("Program"),
        "total_courses": frappe.db.count("Course"),
        "pending_applications": frappe.db.count("Student Applicant", {"application_status": ["in", ["Applied", "Pending"]]}),
        "active_term": frappe.db.get_value(
            "Academic Term",
            {"term_start_date": ["<=", nowdate()], "term_end_date": [">=", nowdate()]},
            "name"
        ),
        "recent_enrollments": frappe.db.count("Student", {"creation": [">=", add_months(nowdate(), -1)]})
    }


def get_hr_dashboard():
    """HR dashboard"""
    return {
        "type": "hr",
        "total_employees": frappe.db.count("Employee", {"status": "Active"}),
        "pending_leaves": frappe.db.count("Leave Application", {"status": "Open"}),
        "open_positions": frappe.db.count("Job Opening", {"status": "Open"}),
        "recent_hires": frappe.db.count("Employee", {"date_of_joining": [">=", add_months(nowdate(), -1)]}),
        "attendance_today": frappe.db.count("Attendance", {"attendance_date": nowdate(), "status": "Present"})
    }


def get_finance_dashboard():
    """Finance dashboard"""
    fee_stats = frappe.db.sql("""
        SELECT
            COALESCE(SUM(grand_total), 0) as total_generated,
            COALESCE(SUM(grand_total - outstanding_amount), 0) as total_collected,
            COALESCE(SUM(outstanding_amount), 0) as total_outstanding
        FROM `tabFees`
        WHERE docstatus = 1
        AND posting_date >= DATE_SUB(CURDATE(), INTERVAL 1 YEAR)
    """, as_dict=True)[0]

    return {
        "type": "finance",
        "total_generated": flt(fee_stats.get("total_generated")),
        "total_collected": flt(fee_stats.get("total_collected")),
        "total_outstanding": flt(fee_stats.get("total_outstanding")),
        "collection_rate": round(
            (fee_stats.get("total_collected") or 0) / (fee_stats.get("total_generated") or 1) * 100, 1
        ),
        "pending_payments": frappe.db.count("Fees", {"docstatus": 1, "outstanding_amount": [">", 0]}),
        "refund_requests": frappe.db.count("Fee Refund", {"docstatus": 0})
    }


def get_default_dashboard():
    """Default dashboard for other users"""
    return {
        "type": "default",
        "welcome_message": _("Welcome to University ERP"),
        "quick_links": [
            {"label": _("Students"), "link": "/app/student"},
            {"label": _("Courses"), "link": "/app/course"},
            {"label": _("Programs"), "link": "/app/program"}
        ]
    }

File: development/frappe-bench/apps/university_erp/university_erp/university_erp/page/role_dashboard/role_dashboard.js
frappe.pages['role-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('My Dashboard'),
        single_column: true
    });

    new RoleDashboard(page);
};

class RoleDashboard {
    constructor(page) {
        this.page = page;
        this.make();
    }

    make() {
        this.$container = $('<div class="role-dashboard p-3"></div>').appendTo(this.page.body);
        this.page.set_primary_action(__('Refresh'), () => this.load_data(), 'refresh');
        this.load_data();
    }

    load_data() {
        this.$container.html(`
            <div class="text-center py-5">
                <div class="spinner-border text-primary" role="status"></div>
                <p class="mt-2 text-muted">${__('Loading your dashboard...')}</p>
            </div>
        `);

        frappe.call({
            method: 'university_erp.university_erp.page.role_dashboard.role_dashboard.get_role_dashboard_data',
            callback: (r) => {
                if (r.message) {
                    this.render(r.message);
                }
            }
        });
    }

    render(data) {
        if (data.error) {
            this.$container.html(`<div class="alert alert-warning">${data.error}</div>`);
            return;
        }

        switch (data.type) {
            case 'student':
                this.render_student_dashboard(data);
                break;
            case 'faculty':
                this.render_faculty_dashboard(data);
                break;
            case 'admin':
                this.render_admin_dashboard(data);
                break;
            case 'hr':
                this.render_hr_dashboard(data);
                break;
            case 'finance':
                this.render_finance_dashboard(data);
                break;
            default:
                this.render_default_dashboard(data);
        }
    }

    render_student_dashboard(data) {
        this.$container.html(`
            <div class="row mb-4">
                <div class="col-12">
                    <h4>${__('Welcome')}, ${data.student_name}</h4>
                    <p class="text-muted">${data.program || ''}</p>
                </div>
            </div>

            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-primary text-white">
                        <i class="fa fa-chart-line"></i>
                        <h3>${data.cgpa || 'N/A'}</h3>
                        <p>${__('CGPA')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-success text-white">
                        <i class="fa fa-calendar-check"></i>
                        <h3>${data.attendance_rate}%</h3>
                        <p>${__('Attendance')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-danger text-white">
                        <i class="fa fa-rupee-sign"></i>
                        <h3>₹${this.formatNumber(data.pending_fees)}</h3>
                        <p>${__('Pending Fees')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-info text-white">
                        <i class="fa fa-book"></i>
                        <h3>${data.books_borrowed}</h3>
                        <p>${__('Books Borrowed')}</p>
                    </div>
                </div>
            </div>

            <div class="row">
                <div class="col-md-6 mb-3">
                    <div class="card">
                        <div class="card-header"><h6>${__('My Courses')}</h6></div>
                        <div class="card-body">
                            ${data.enrollments.length ? data.enrollments.map(e => `
                                <div class="d-flex justify-content-between border-bottom py-2">
                                    <span>${e.course_name || e.course}</span>
                                </div>
                            `).join('') : `<p class="text-muted">${__('No courses enrolled')}</p>`}
                        </div>
                    </div>
                </div>
                <div class="col-md-6 mb-3">
                    <div class="card">
                        <div class="card-header"><h6>${__('Recent Results')}</h6></div>
                        <div class="card-body">
                            ${data.latest_results.length ? data.latest_results.map(r => `
                                <div class="d-flex justify-content-between border-bottom py-2">
                                    <span>${r.course_name || r.course}</span>
                                    <span class="badge bg-primary">${r.grade || 'N/A'}</span>
                                </div>
                            `).join('') : `<p class="text-muted">${__('No results available')}</p>`}
                        </div>
                    </div>
                </div>
            </div>

            <div class="row mt-3">
                <div class="col-12">
                    <div class="card">
                        <div class="card-header"><h6>${__('Quick Actions')}</h6></div>
                        <div class="card-body">
                            <a href="/app/fees?student=${data.student_name}" class="btn btn-outline-primary me-2 mb-2">${__('View Fees')}</a>
                            <a href="/placement" class="btn btn-outline-success me-2 mb-2">${__('Placement Portal')}</a>
                            <a href="/library" class="btn btn-outline-info me-2 mb-2">${__('Library')}</a>
                        </div>
                    </div>
                </div>
            </div>
        `);
    }

    render_faculty_dashboard(data) {
        this.$container.html(`
            <div class="row mb-4">
                <div class="col-12">
                    <h4>${__('Welcome')}, ${data.employee_name}</h4>
                    <p class="text-muted">${data.department || ''}</p>
                </div>
            </div>

            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-primary text-white">
                        <i class="fa fa-book"></i>
                        <h3>${data.assignments.length}</h3>
                        <p>${__('Assigned Courses')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-success text-white">
                        <i class="fa fa-clock"></i>
                        <h3>${data.total_hours}</h3>
                        <p>${__('Weekly Hours')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-warning text-white">
                        <i class="fa fa-star"></i>
                        <h3>${data.avg_feedback}/5</h3>
                        <p>${__('Avg Feedback')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-info text-white">
                        <i class="fa fa-calendar"></i>
                        <h3>${data.pending_leaves}</h3>
                        <p>${__('Pending Leaves')}</p>
                    </div>
                </div>
            </div>

            <div class="row">
                <div class="col-md-6 mb-3">
                    <div class="card">
                        <div class="card-header"><h6>${__("Today's Schedule")}</h6></div>
                        <div class="card-body">
                            ${data.today_schedule.length ? data.today_schedule.map(s => `
                                <div class="d-flex justify-content-between border-bottom py-2">
                                    <span>${s.course_name}</span>
                                    <span class="text-muted">${s.start_time} - ${s.end_time}</span>
                                </div>
                            `).join('') : `<p class="text-muted">${__('No classes today')}</p>`}
                        </div>
                    </div>
                </div>
                <div class="col-md-6 mb-3">
                    <div class="card">
                        <div class="card-header"><h6>${__('Leave Balance')}</h6></div>
                        <div class="card-body">
                            ${data.leave_balance.length ? data.leave_balance.map(l => `
                                <div class="d-flex justify-content-between border-bottom py-2">
                                    <span>${l.leave_type}</span>
                                    <span>${l.total_leaves_allocated - l.total_leaves_taken} / ${l.total_leaves_allocated}</span>
                                </div>
                            `).join('') : `<p class="text-muted">${__('No leave allocation')}</p>`}
                        </div>
                    </div>
                </div>
            </div>
        `);
    }

    render_admin_dashboard(data) {
        this.$container.html(`
            <div class="row mb-4">
                <div class="col-12">
                    <h4>${__('Academic Administration Dashboard')}</h4>
                    <p class="text-muted">${__('Current Term')}: ${data.active_term || 'N/A'}</p>
                </div>
            </div>

            <div class="row mb-4">
                <div class="col-md-2 col-sm-6 mb-3">
                    <div class="stat-card bg-primary text-white">
                        <h3>${data.total_students}</h3>
                        <p>${__('Students')}</p>
                    </div>
                </div>
                <div class="col-md-2 col-sm-6 mb-3">
                    <div class="stat-card bg-success text-white">
                        <h3>${data.total_programs}</h3>
                        <p>${__('Programs')}</p>
                    </div>
                </div>
                <div class="col-md-2 col-sm-6 mb-3">
                    <div class="stat-card bg-info text-white">
                        <h3>${data.total_courses}</h3>
                        <p>${__('Courses')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-warning text-white">
                        <h3>${data.pending_applications}</h3>
                        <p>${__('Pending Applications')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-secondary text-white">
                        <h3>${data.recent_enrollments}</h3>
                        <p>${__('New Enrollments')}</p>
                    </div>
                </div>
            </div>
        `);
    }

    render_hr_dashboard(data) {
        this.$container.html(`
            <div class="row mb-4">
                <div class="col-12"><h4>${__('HR Dashboard')}</h4></div>
            </div>
            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-primary text-white">
                        <h3>${data.total_employees}</h3>
                        <p>${__('Employees')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-warning text-white">
                        <h3>${data.pending_leaves}</h3>
                        <p>${__('Pending Leaves')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-success text-white">
                        <h3>${data.attendance_today}</h3>
                        <p>${__('Present Today')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-info text-white">
                        <h3>${data.recent_hires}</h3>
                        <p>${__('New Hires')}</p>
                    </div>
                </div>
            </div>
        `);
    }

    render_finance_dashboard(data) {
        this.$container.html(`
            <div class="row mb-4">
                <div class="col-12"><h4>${__('Finance Dashboard')}</h4></div>
            </div>
            <div class="row mb-4">
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-primary text-white">
                        <h3>₹${this.formatNumber(data.total_generated)}</h3>
                        <p>${__('Generated')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-success text-white">
                        <h3>₹${this.formatNumber(data.total_collected)}</h3>
                        <p>${__('Collected')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-danger text-white">
                        <h3>₹${this.formatNumber(data.total_outstanding)}</h3>
                        <p>${__('Outstanding')}</p>
                    </div>
                </div>
                <div class="col-md-3 col-sm-6 mb-3">
                    <div class="stat-card bg-info text-white">
                        <h3>${data.collection_rate}%</h3>
                        <p>${__('Collection Rate')}</p>
                    </div>
                </div>
            </div>
        `);
    }

    render_default_dashboard(data) {
        this.$container.html(`
            <div class="text-center py-5">
                <h3>${data.welcome_message}</h3>
                <div class="mt-4">
                    ${data.quick_links.map(l => `
                        <a href="${l.link}" class="btn btn-outline-primary m-2">${l.label}</a>
                    `).join('')}
                </div>
            </div>
        `);
    }

    formatNumber(value) {
        if (!value) return '0';
        if (value >= 10000000) return (value / 10000000).toFixed(1) + ' Cr';
        if (value >= 100000) return (value / 100000).toFixed(1) + ' L';
        if (value >= 1000) return (value / 1000).toFixed(1) + ' K';
        return value.toFixed(0);
    }
}
